<?xml version="1.0" encoding="utf-8"?>
<launch>
<!-- This file launches the pointcloud2_interpreter as a nodelet and the laserscan_interpreter as a node.
     Use the arguments defined below to control what is launched and whether the recorded bag file should be played. -->
  
  
  <!-- uncomment the following two lines to set logger level to DEBUG for find_moving_objects -->
<!--   <env name="ROSCONSOLE_CONFIG_FILE" -->
<!--        value="$(find find_moving_objects)/logger_config/logger_config_debug.config" /> -->
  
  
  <!-- Command line arguments -->
  <arg name="use_lidar"
       default="true"
       doc="Set to false if lidar should not be used"
       />
  
  <arg name="use_camera"
       default="false"
       doc="Set to false if camera should not be used"
       />
  
  <arg name="downscale_pointcloud"
       default="false"
       doc="Set to true if voxel grid filter should be applied to camera output"
       />
  
  <arg name="laserscan_interpreter_extra_args"
       default=""
       doc="Additional args for the laserscan interpreter"
       />
  
  <arg name="pointcloud2_interpreter_extra_args"
       default=""
       doc="Additional args for the pointcloud2 interpreter"
       />
  
  <arg name="frame_broadcaster_extra_args"
       default=""
       doc="Additional args for the frame broadcaster"
       />

  
  <!-- Frames broadcaster - tf state publisher -->
  <node pkg="find_moving_objects"
        type="example_frame_broadcaster_node"
        name="frame_broadcaster"
        ns="frame_broadcaster"
        args="--print_all_options
              $(arg frame_broadcaster_extra_args)"
        output="screen"
        />

  <!--  Lidar  -->
  <group if="$(arg use_lidar)"  ns="lidar">
    <!-- Run laserscan interpreter for lidar -->
    <include file="$(find find_moving_objects)/launch/includes/laserscan_interpreter_node.launch.xml">
      <arg unless="$(arg use_bag)" 
           name="subscribe_topic" value="scan" />
      <arg if="$(arg use_bag)" 
           name="subscribe_topic" value="scan_echoed" />
      <arg name="topic_objects" value="/moving_objects" />
      <arg name="subscribe_buffer_size" value="10" />
    </include>
  </group>
  
  <!--  Camera  -->   
  <!--  Attach to camera's namespace  -->
  <group if="$(arg use_camera)"  ns="camera">
    <!-- Voxel filter to down-scale pointcloud2 -->
    <node if="$(eval arg('downscale_pointcloud') and not arg('use_bag'))"
          pkg="nodelet" 
          type="nodelet" 
          name="voxel_grid" 
          args="load pcl/VoxelGrid realsense2_camera_manager" 
          output="screen">
      <remap from="~input" to="depth/color/points" />
      <remap from="~output" to="depth/color/points/filtered" />
      <rosparam>
        filter_field_name: z
        filter_limit_min: 0.01
        filter_limit_max: 6.5
        filter_limit_negative: False
        leaf_size: 0.01
  <!--       output_frame: camera_link -->
      </rosparam>
    </node>
    
    <!-- Run pointcloud2 interpreter for camera -->
<!--     <include file="$(find find_moving_objects)/launch/includes/pointcloud2_interpreter_node.launch.xml"> -->
    <include file="$(find find_moving_objects)/launch/includes/pointcloud2_interpreter_nodelet.launch.xml">
      <arg name="manager" value="realsense2_camera_manager" />
      <arg if="$(arg use_bag)"
           name="create_manager" value="true" />
      <arg if="$(arg use_bag)" 
           name="subscribe_topic" value="depth/color/points/filtered_echoed" />
      <arg if="$(eval not arg('downscale_pointcloud') and not arg('use_bag'))" 
           name="subscribe_topic" value="depth/color/points" />
      <arg if="$(eval arg('downscale_pointcloud') and not arg('use_bag'))" 
           name="subscribe_topic" value="depth/color/points/filtered" />
      <arg name="topic_objects" value="/moving_objects" />
      <arg name="publish_ema" value="false"/>
      <arg name="publish_objects_closest_points_markers" value="false"/>
      <arg name="subscribe_buffer_size" value="10" />
    </include>
</group>
  
  <!-- Run confidence enhancing node -->
  <node pkg="find_moving_objects"
        type="moving_objects_confidence_enhancer_node"
        name="confidence_enhancer_node"
        ns="confidence_enhancer"
        output="screen" >
    <param name="subscribe_topic" value="/moving_objects" />
  </node>
</launch>
